### PR TITLE
[MI-2460] Fix security issues

### PR DESCRIPTION
**Pull request by ayusht2810**
Fix security issues:
https://github.com/mattermost/mattermost-plugin-servicenow/issues/144
https://github.com/mattermost/mattermost-plugin-servicenow/issues/146
https://github.com/mattermost/mattermost-plugin-servicenow/issues/147
https://github.com/mattermost/mattermost-plugin-servicenow/issues/148
https://github.com/mattermost/mattermost-plugin-servicenow/issues/149
https://github.com/mattermost/mattermost-plugin-servicenow/issues/150

### CODE REVIEW

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/plugin/api.go [hunk 1]
```diff
@@ -212,6 +227,24 @@ func (p *Plugin) createSubscription(w http.ResponseWriter, r *http.Request) {
...
+	userID := r.Header.Get(constants.HeaderMattermostUserID)
+	if userID != *subscription.UserID {
+		p.API.LogError(constants.ErrorUserMismatch)
+		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusBadRequest, Message: constants.ErrorUserMismatch})
+		return
+	}
+
+	hasPermission, permissionErr := p.HasChannelPermissions(userID, *subscription.ChannelID)
+	if permissionErr != nil {
+		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusInternalServerError, Message: permissionErr.Error()})
```
[the remark this answers is not in the document]
Comment (ayusht2810) [reply server/plugin/api.go:239]: it is already been done in the function `HasChannelPermissions`

**Review by raghavaggarwal2308 — APPROVED** [review]

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/api_test.go [hunk 2]
```diff
@@ -1151,10 +1373,42 @@ func TestGetAllSubscriptions(t *testing.T) {
...
+			SetupPlugin: func(p *Plugin) {
+				monkey.PatchInstanceMethod(reflect.TypeOf(p), "HasChannelPermissions", func(_ *Plugin, _, _ string) (bool, error) {
+					return false, fmt.Errorf(constants.ErrorChannelPermissionsForUser)
+				})
+			},
+			ExpectedStatusCode:   http.StatusInternalServerError,
+			ExpectedErrorMessage: constants.ErrorChannelPermissionsForUser,
+			ExpectedCount:        -1,
+		},
+		"user do not have the permissions for the subscriptions channel": {
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1397]:
```suggestion
		"user does not have permission for the subscriptions channel": {
```

File: server/plugin/api_test.go [hunk 3]
```diff
@@ -1306,11 +1572,55 @@ func TestEditSubscription(t *testing.T) {
...
+			},
+			SetupPlugin: func(p *Plugin) {
+				monkey.PatchInstanceMethod(reflect.TypeOf(p), "HasChannelPermissions", func(_ *Plugin, _, _ string) (bool, error) {
+					return false, fmt.Errorf(constants.ErrorChannelPermissionsForUser)
+				})
+			},
+			ExpectedStatusCode:   http.StatusInternalServerError,
+			ExpectedErrorMessage: constants.ErrorChannelPermissionsForUser,
+		},
+		"user do not have the permission to set subscription for this channel": {
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1599]:
```suggestion
		"user does not have permission to set the subscription for this channel": {
```

File: server/plugin/command_test.go [hunk 4]
```diff
@@ -557,10 +562,57 @@ func TestHandleListSubscriptions(t *testing.T) {
...
+					return true, nil
+				})
+			},
 			isResponse:       true,
 			expectedResponse: fmt.Sprintf("#### Bulk subscriptions\n| Subscription ID | Record Type | Events | Created By | Channel |\n| :----|:--------| :--------|:--------|:--------|\n|%s|Problem|Priority changed, State changed|N/A|N/A|\n#### Record subscriptions\n| Subscription ID | Record Type | Record Number | Record Short Description | Events | Created By | Channel |\n| :----|:--------| :--------| :-----| :--------|:--------|:--------|\n|%s|Problem|PRB0000005|Test description|Priority changed, State changed|N/A|N/A|", testutils.GetServiceNowSysID(), testutils.GetServiceNowSysID()),
 			expectedError:    listSubscriptionsWaitMessage,
 		},
+		{
+			description: "HandleListSubscriptions: Unable to get permissions for channel",
+			params:      []string{"me", "all_channels"},
```
Comment (Kshitij-Katiyar) [inline server/plugin/command_test.go:576]: are there constant for these

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/plugin/http.go [hunk 5]
```diff
@@ -69,7 +69,16 @@ func (c *client) Call(method, path, contentType string, inBody io.Reader, out in
 		if strings.Contains(err.Error(), "invalid character '<'") {
 			return nil, http.StatusInternalServerError, ErrorContentTypeNotJSON
 		}
-		return nil, http.StatusInternalServerError, err
+
+		updatedError := strings.ReplaceAll(err.Error(), c.plugin.getConfiguration().MattermostSiteURL, "")
```
Comment (manojmalik20) [inline server/plugin/http.go:73]:
```suggestion
		updatedError := strings.ReplaceAll(err.Error(), c.plugin.getConfiguration().ServiceNowBaseURL, "")
```
We have to replace the ServiceNow URL as we are making request to that URL.

File: server/plugin/util.go [hunk 6]
```diff
@@ -308,23 +308,23 @@ func decodeKey(key string) (string, error) {
 	return string(decodedKey), nil
 }
 
-func (p *Plugin) HasChannelPermissions(userID, channelID string) (bool, error) {
+func (p *Plugin) HasChannelPermissions(userID, channelID string) (bool, int, error) {
```
Comment (manojmalik20) [inline server/plugin/util.go:311]: Is there really a need for the `bool` return type?

File: server/plugin/api_test.go [hunk 7]
```diff
@@ -446,7 +479,76 @@ func TestShareRecordInChannel(t *testing.T) {
...
+			SetupClient: func(client *mock_plugin.Client) {},
+			SetupPlugin: func(p *Plugin) {
+				monkey.PatchInstanceMethod(reflect.TypeOf(p), "HasChannelPermissions", func(_ *Plugin, _, _ string) (int, error) {
+					return http.StatusInternalServerError, fmt.Errorf(constants.ErrorChannelPermissionsForUser)
+				})
+			},
+			ExpectedErrorMessage: constants.ErrorChannelPermissionsForUser,
+			ExpectedStatusCode:   http.StatusInternalServerError,
+		},
+		"user do not have the permission to share record in the channel": {
```
Comment (manojmalik20) [inline server/plugin/api_test.go:507]:
```suggestion
		"user does not have the permission to share record in the channel": {
```

File: server/plugin/api_test.go [hunk 8]
```diff
@@ -994,11 +1119,77 @@ func TestCreateSubscription(t *testing.T) {
...
+			},
+			SetupPlugin: func(p *Plugin) {
+				monkey.PatchInstanceMethod(reflect.TypeOf(p), "HasChannelPermissions", func(_ *Plugin, _, _ string) (int, error) {
+					return http.StatusInternalServerError, fmt.Errorf(constants.ErrorChannelPermissionsForUser)
+				})
+			},
+			ExpectedStatusCode:   http.StatusInternalServerError,
+			ExpectedErrorMessage: constants.ErrorChannelPermissionsForUser,
+		},
+		"user do not have the permission to create subscription for this channel": {
```
Comment (manojmalik20) [inline server/plugin/api_test.go:1168]:
```suggestion
		"user does not have permission to create a subscription for this channel": {
```

File: server/plugin/api_test.go [hunk 9]
```diff
@@ -1028,11 +1227,19 @@ func TestCreateSubscription(t *testing.T) {
...
+			},
 			ExpectedStatusCode:   http.StatusBadRequest,
 			ExpectedErrorMessage: "Subscription already exists",
 		},
 		"failed to create subscription": {
-			RequestBody: "{}",
+			RequestBody: fmt.Sprintf(`{
+				"user_id": "%s",
+				"channel_id": "%s"
+			  	}`, testutils.GetID(), testutils.GetChannelID()),
```
Comment (manojmalik20) [inline server/plugin/api_test.go:1242]: Why have you changed this everywhere?

File: server/plugin/api_test.go [hunk 10]
```diff
@@ -1306,11 +1571,55 @@ func TestEditSubscription(t *testing.T) {
...
+			SetupAPI: func(api *plugintest.API) {},
+			SetupClient: func(client *mock_plugin.Client) {
+				var s *serializer.SubscriptionPayload
+				monkey.PatchInstanceMethod(reflect.TypeOf(s), "IsValidForUpdation", func(_ *serializer.SubscriptionPayload, _ string) error {
+					return nil
+				})
```
Comment (manojmalik20) [inline server/plugin/api_test.go:1588]: Why is this inside `SetupClient`? Shouldn't it be inside `SetupPlugin`? 
Fix this everywhere.

File: server/plugin/api_test.go [hunk 11]
```diff
@@ -1306,11 +1571,55 @@ func TestEditSubscription(t *testing.T) {
...
+			},
+			SetupPlugin: func(p *Plugin) {
+				monkey.PatchInstanceMethod(reflect.TypeOf(p), "HasChannelPermissions", func(_ *Plugin, _, _ string) (int, error) {
+					return http.StatusInternalServerError, fmt.Errorf(constants.ErrorChannelPermissionsForUser)
+				})
+			},
+			ExpectedStatusCode:   http.StatusInternalServerError,
+			ExpectedErrorMessage: constants.ErrorChannelPermissionsForUser,
+		},
+		"user does not have permission to set the subscription for this channel": {
```
Comment (manojmalik20) [inline server/plugin/api_test.go:1598]:
```suggestion
		"user does not have permission to edit the subscription for this channel": {
```

File: server/plugin/command_test.go [hunk 12]
```diff
@@ -557,13 +563,69 @@ func TestHandleListSubscriptions(t *testing.T) {
...
+				monkey.PatchInstanceMethod(reflect.TypeOf(p), "HasChannelPermissions", func(_ *Plugin, _, _ string) (int, error) {
+					return http.StatusInternalServerError, fmt.Errorf(constants.ErrorChannelPermissionsForUser)
+				})
+			},
+			isResponse:       true,
+			expectedResponse: constants.ErrorNoActiveSubscriptions,
+			expectedError:    listSubscriptionsWaitMessage,
+		},
+		{
+			description: "HandleListSubscriptions: User do not have permissions for the subscriptions channel",
```
Comment (manojmalik20) [inline server/plugin/command_test.go:601]:
```suggestion
			description: "HandleListSubscriptions: User does not have permissions for the subscriptions in the channel",
```

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**